internal/validitywindow/handler_test.go

```diff
@@ -0,0 +1,316 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package validitywindow
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBlockFetcherHandler_FetchBlocks(t *testing.T) {
+	tests := []struct {
+		name         string
+		setupBlocks  func() map[uint64]ExecutionBlock[container]
+		blockHeight  uint64
+		minTimestamp int64
+		wantErr      bool
+		wantBlocks   int
+	}{
+		{
+			name: "happy path - fetch all blocks",
+			setupBlocks: func() map[uint64]ExecutionBlock[container] {
+				return generateBlockChain(10, 3)
+			},
+			blockHeight:  9, // The tip of the chain
+			minTimestamp: 3, // Should get blocks with timestamp of 3 - 1 = 2 where 2 is the boundary block
+			wantBlocks:   8,
+		},
+		{
+			name: "partial response",
+			setupBlocks: func() map[uint64]ExecutionBlock[container] {
+				blocks := generateBlockChain(10, 3)
+				delete(blocks, uint64(7))
+				return blocks
+			},
+			blockHeight:  9,
+			minTimestamp: 0,
+			wantBlocks:   2, // Should get block 9 and 8 before failing on 7
+		},
+		{
+			name:         "no blocks - should error",
+			setupBlocks:  func() map[uint64]ExecutionBlock[container] { return nil },
+			blockHeight:  9,
+			minTimestamp: 3,
+			wantBlocks:   0,
+			wantErr:      true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			r := require.New(t)
+
+			blocks := tt.setupBlocks()
+			retriever := newTestBlockRetriever(withBlocks(blocks))
+
+			handler := NewBlockFetcherHandler(retriever)
+			fetchedBlocks, err := handler.fetchBlocks(ctx, &BlockFetchRequest{
+				BlockHeight:  tt.blockHeight,
+				MinTimestamp: tt.minTimestamp,
+			})
+
+			if tt.wantErr {
+				r.Error(err)
+			} else {
+				r.NoError(err)
+			}
+
+			r.Len(fetchedBlocks, tt.wantBlocks)
+			for _, blk := range fetchedBlocks {
+				block, err := retriever.parseBlock(blk)
+				r.NoError(err)
+
+				_, ok := blocks[block.GetHeight()]
+				r.True(ok)
+			}
```

Comment:
Can we make blocks parseable and check the block we receive matches the block we expect?